open_widget_framework/open_widget_framework/views.py

```diff
@@ -1,204 +1,172 @@
-        return WidgetListView.get(request, widget_list_id)
+        if api_settings.WIDGET_LIST_EDIT_PERMISSIONS:
+            widget_list = get_object_or_404(WidgetList, pk=self.kwargs['widget_list_id'])
+            if not self.request.user.has_perms(api_settings.WIDGET_LIST_EDIT_PERMISSIONS, widget_list):
+                #TODO Handle permissions denied
+                self.permission_denied(self.request, "This user does not have permission to edit that widget list")
 
-    def put(self, request, widget_list_id, widget_id):
+    def get_queryset(self):
         """
-        API endpoint to update the data for a widget instance
+        get_queryset returns all widgets belonging to the list specified in kwargs['widget_list_id']
+        """
+        return get_object_or_404(WidgetList, pk=self.kwargs['widget_list_id']).get_widgets()
 
-        Args:
-            widget_list_id (int): id of WidgetList containing the Widget
-            widget_id (int): id of Widget to update
+    def retrieve(self, request, *args, **kwargs):
+        """
+        retrieve is used for acquiring data to update a widget. It returns the configuration for form rendering and the
+            widgetData for the widget specified
         """
-        widget_list = get_object_or_404(WidgetList, id=widget_list_id)
-        # TODO: widget_list.can_access
+        serializer = self.serializer_class(self.get_object())
+        return JsonResponse({
+            'widgetClassConfigurations': {
+                serializer.data['widget_class']: get_widget_class_configurations()[serializer.data['widget_class']],
+            },
+            'widgetData': serializer.get_form_data(),
+        })
 
-        widget = get_object_or_404(WidgetInstance, id=widget_id)
-        update_data = loads(request.body.decode())
+    def create(self, request, *args, **kwargs):
+        """
+        API endpoint to create a widget instance on a list after validating the data with the serializer class.
+            Returns the updated widget-list
+        """
+        self.check_widget_list_edit_permissions()
+        super().create(request, *args, **kwargs)
+        return make_widget_list_response(self.get_queryset())
 
-        # validate the data using the widget serializer class and then update
-        serializer = WidgetInstance.get_widget_class_serializer(
-            update_data.pop("widget_class")
-        )(data=update_data)
-        if serializer.is_valid():
-            widget.title = update_data.pop("title")
-            widget.configuration = update_data
+    def destroy(self, request, *args, **kwargs):
+        """
+        API endpoint to delete a widget instance from a list and reposition the remaining widgets on the list.
+            Returns the updated widget list
+        """
+        self.check_widget_list_edit_permissions()
+        widget_to_delete = self.get_object()
+        for widget in self.get_queryset().filter(position__gt=widget_to_delete.position):
```

Comment:
Similar problem to the below issue. I would `select_for_update()` all widgets on this list in order to lock them and guarantee integrity of the ordering as you update them.